dds/idl/be_global.cpp

```diff
@@ -602,3 +619,93 @@ BE_GlobalData::get_include_block(BE_GlobalData::stream_enum_t which)
 
   return ret.str();
 }
+
+void
+BE_GlobalData::cache_annotations()
+{
+  if (idl_global->idl_version_ >= IDL_VERSION_4) {
+    UTL_Scope* root = idl_global->scopes().bottom();
+    topic_annotation_ = dynamic_cast<AST_Annotation_Decl*>(
+      root->lookup_by_name("::@topic"));
+    key_annotation_ = dynamic_cast<AST_Annotation_Decl*>(
+      root->lookup_by_name("::@key"));
+    nested_annotation_ = dynamic_cast<AST_Annotation_Decl*>(
+      root->lookup_by_name("::@nested"));
+    default_nested_annotation_ = dynamic_cast<AST_Annotation_Decl*>(
+      root->lookup_by_name("::@default_nested"));
+  }
+}
+
+bool
+BE_GlobalData::is_topic_type(AST_Decl* node)
+{
+  /* A type is a topic type and gets type support if it is annotated with
+   * @topic or else is not nested.
+   */
+  return node->annotations().find(topic_annotation_) || !is_nested(node);
+}
+
+bool
+BE_GlobalData::is_nested(AST_Decl* node)
+{
+  // If annotated with @nested, then @nested.value is the type's nested value.
+  AST_Annotation_Appl* nested = node->annotations().find(nested_annotation_);
+  if (nested) {
+    return dynamic_cast<AST_Annotation_Member*>((*nested)["value"])->
```

Comment:
This is related to [the conversation I already had with Tim](https://github.com/objectcomputing/OpenDDS/pull/1067#discussion_r296289905). The issue is that TAO_IDL uses pointers heavily, but I don't think any of these things would fail unless someone messed up the AST somehow. However covariant return types is new to me, but makes sense and I agree would be perfect for this situation.